module-apps/ModuleApps.md

```diff
@@ -1,10 +1,179 @@
 Module apps
 ===========
 
+# What's an application
+
+In MuditaOS application is a service derivative which:
+- Have and have to have access to UI 
+- Have limited lifespan
+- Can operate with actions and popups
+
+# How to add application
+
+This paragraph is using `applicaton-test` as an example.
+
+**WARNING** All applications are enabled by default. Please disable yours if it's not always needed in [CmakeLists.txt](CMakeLists.txt). I.e. following: `option(ENABLE_APP_TEST "Enable application antenna" OFF)`
+
+* Add new `application` catalog in `module-apps` i.e. `application-test`
+* Add a CMake for it, simple example here: [application-test/CMakeLists.txt](application-test/CMakeLists.txt)
+* Create:
+    * include/application-test/ for public includes for this application
+    * create appliation header with:
+        * Appilcation definition - inheriting from `Application`
+        * AppliactionManifest - stating what actions can this application support, and it's lock policy
+**NOTE** We strongly encurage using [MVP](https://en.wikipedia.org/wiki/Model–view–presenter) application GUI pattern.
+* create presenters & windows catalogs with all the windows and presenters you require. Each application require at least one window.
+* Write `InitHandler()` for application.
+    * **NOTE** This is your application entry point!
+    * **NOTE** if it's not copied properly - application won't start and work
+Please see example in: [ApplicationTest.cpp](application-test/ApplicationTest.cpp) for `InitHandler()`
+* Write `createUserInterface()`, which role is:
+    * provide windowsFactory with windows blueprints for this application
+    * inform what popups this application handles
+**WARNING** default window for each application is always named `gui::name::window::main_window` without said window application may be never shown properly
+
+Basic example can be found in this catalog: [application-test](./application-test)
+
+# How to launch app:
+
+1. From service/app via action `Launch`
+example:
+``` c++
+return app::manager::Controller::sendAction(application,
+                                            app::manager::actions::Launch,
+                                            std::make_unique<app::ApplicationLaunchData>("ApplicationAlarmClock"));
+                                  });
+
+```
+2. Enable it in background with parameter: `startInBackground` in Application constructor
+
+**WARNING** Enabling app in background means it will always lay on stack and heap and never end
+**WARNING** App in background just doesn't handle `Launch` this means that if you call `Launch` on the system, there will be [Terrible, terrible damage](https://www.youtube.com/watch?v=YKZlcgOzi_E)
```

Comment:
I'll split the response here into two parts:

1. Log part: 

I was wondering if it will get through :D 
In the system, in ApplicationCommon.cpp there is a log: 
```
LOG_FATAL("!!! Terrible terrible damage! application with no focus grabbed key!");
```
that actually is a reference to these SC2 casts :) After some major reworks, I was a tad frustrated about the fact that only 2/3rds of the mess in application launching was fixed - as the only thing I really required at the moment was to have some application stack. The part on apps <-> app manager <-> event manager synchronization was left out.
As you can see - issues known years ago are still in place :) 

Saying that - I'll reluctantly remove the SC2 reference ;) 

2. ` not send Launch action to this app`

This is a tricky question. Short answer: `you should send launch to the app to start`. You should actually send the launch to any app to start except the home one.
The problem is with the synchronization part 
- we can send launch action before we have the application manager ready to launch applications. 
- there is no way to synchronize dependency to application manager from other services
This clearly is a bug in the software. We do not see it daily most probably because system startup is slow enough.

Reply:
Im in favor of small easter eggs in code and documentation, especially in opensource. Its like a smile from the developer for future readers to thank them for spending time on the code :)